src/components/Notifier/Notifier.tsx

```diff
@@ -0,0 +1,237 @@
+  seed += 1;
+  return `notifier_${now}_${id}`;
+}
+
+// Notifier is wrapping notices in a TransitionGroup
+const Notifier = defineComponent<NotifierProps>({
+  name: 'Notifier',
+  inheritAttrs: false,
+  props: ['prefixCls', 'transitionName', 'animation', 'maxCount', 'closeIcon'] as any,
+  setup(props, { attrs, expose, slots }) {
+
+    const hookRefs = new Map<Key, HTMLDivElement>();
+
+    // ref containing all currently displayed notices
+    const notices = ref<NotifierState>([]);
+
+    const transitionProps = computed(() => {
+      const { prefixCls, animation = 'fade' } = props;
+      let name = props.transitionName;
+
+      if (!name && animation) {
+        name = `${prefixCls}_${animation}`;
+      }
+
+      return getTransitionGroupProps(name!);
+    });
+
+    // method we expose to add a notice to the notice list
+    const add = (originNotice: NoticeContent, holderCallback?: HolderReadyCallback) => {
+      const key = originNotice.key || getUuid();
+      const notice: NoticeContent & { key: Key; userPassKey?: Key } = {
+        ...originNotice,
+        key,
+      };
+      const { maxCount } = props;
+      const noticeIndex = notices.value.map(v => v.notice.key).indexOf(key);
+      const updatedNotices = notices.value.concat();
+
+      if (noticeIndex !== -1) {
+        updatedNotices.splice(noticeIndex, 1, { notice, holderCallback } as any);
+      } else {
+        if (maxCount && notices.value.length >= maxCount) {
+          notice.key = updatedNotices[0].notice.key as Key;
+          notice.updateMark = getUuid();
+          notice.userPassKey = key;
+          updatedNotices.shift();
+        }
+        updatedNotices.push({ notice, holderCallback } as any);
+      }
+      notices.value = updatedNotices;
+    };
+
+    // method we expose to remove a notice to the notice list
+    const remove = (removeKey: Key) => {
+      notices.value = notices.value.filter(({ notice: { key, userPassKey } }) => {
+        const mergedKey = userPassKey || key;
+        return mergedKey !== removeKey;
+      });
+    };
+
+    expose({
+      add,
+      remove,
+      notices,
+    });
+
+    return () => {
+      const { prefixCls, closeIcon = slots.closeIcon?.({ prefixCls }) } = props;
+
+      const noticeNodes = notices.value.map(({ notice, holderCallback }, index) => {
+        const updateMark = index === notices.value.length - 1 ? notice.updateMark : undefined;
+        const { key, userPassKey } = notice;
+
+        const { content } = notice;
+
+        // @ts-ignore
+        const noticeProps = {
+          prefixCls,
+          closeIcon: typeof closeIcon === 'function' ? closeIcon({ prefixCls }) : closeIcon,
+          ...(notice as any),
+          ...notice.props,
+          key,
+          noticeKey: userPassKey || key,
+          updateMark,
+          onClose: (noticeKey: Key) => {
+            remove(noticeKey);
+            notice.onClose?.();
+          },
+          onClick: notice.onClick,
+        };
```

Comment:
Type instantiation is excessively deep and possibly infinite.

Reply:
yea typescript has a bad type system

Reply:
🤡 nice